DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs

```diff
@@ -48,7 +48,31 @@ CharEncoding CharEncoding
         UnmanagedType UnmanagedArraySubType,
         int ArraySizeConst,
         short ArraySizeParamIndex,
-        CharEncoding CharEncoding) : MarshallingInfoStringSupport(CharEncoding);
+        CharEncoding CharEncoding) : MarshallingInfoStringSupport(CharEncoding)
+    {
+        public MarshalAsInfo(UnmanagedType unmanagedType, CharEncoding charEncoding)
+            :this(unmanagedType, 0, 0, 0, charEncoding)
+        {
+        }
+
+        public MarshallingInfo CreateArraySubTypeMarshalAsInfo()
+        {
+            Debug.Assert(UnmanagedType is UnmanagedType.LPArray or UnmanagedType.ByValArray or UnmanagedType.SafeArray);
+            if (UnmanagedArraySubType == 0)
+            {
+                return NoMarshallingInfo.Instance;
+            }
+            return new MarshalAsInfo(UnmanagedArraySubType, CharEncoding);
+        }
+
+        /// <summary>
+        /// Helper method to enable cleaner pattern matching for the common case of
+        /// a MarshalAs attribute that just uses the constructor parameter and no additional properties.
+        /// </summary>
+        public void Deconstruct(out UnmanagedType unmanagedType) => unmanagedType = UnmanagedType;
```

Comment:
This doesn't seem to be used anywhere.

Reply:
The `Deconstruct` method is  a special method name for pattern matching. In this case, it enables the `x is MarshalAsInfo(UnmanagedType.I4)` syntax instead of the `x is MarshalAsInfo { UnmanagedType: UnmanagedType.I4 }`. The constructor style syntax is much cleaner in my opinion.

Reply:
Now that we have the two-parameter constructor with the CharSet however, I might modify this method to instead mirror the second constructor with the CharSet parameter.

Reply:
When/how often is this called? A very common complaint about C++ is implicit constructors and this seems very similar in context. I personally am not fan of the current switch statement we have because it is no longer a jump table but rather a much more complex comparison matrix. @elinor-fung removed the `char` and `string` because it was unnecessarily complex and this seems to add to enable that pattern which is far too dense in my experience What are we gaining with this pattern? I am close to saying this isn't an approach we should be using given it is much clearer to have specialized functions that are easier to maintain than the enormous switch statement.

Is there an argument for this approach?

Reply:
The main argument for adding the `Deconstruct` method is to make all of the patterns that were of the form `MarshalAsInfo { UnmanagedType: UnmanagedType.I4 }` to instead be of the form `MarshalAsInfo(UnmanagedType.I4)`. In our marshaller selector in this PR, it's being used 15 times. This syntax also enables the simple MarshalAs checks that just depend on the first parameter to the MarshalAs attribute to look like the application of the MarshalAs attribute, which is a nice touch.

If I move some of the properties (such as the array-specific ones) on the MarshalAs record from being positional (in the constructor) to just regular init properties, I can remove the need for this Deconstruct method by making the two-parameter version be auto-generated. I'll try a few things out to make the Deconstruct method unnecessary.

Reply:
Appreciate the effort.